cumulus/client/cirrus-executor/src/worker.rs

```diff
@@ -255,37 +252,52 @@ async fn handle_block_import_notifications<
             {
                 tracing::error!(
                     target: LOG_TARGET,
-                    "Collation generation processing error: {error}"
+                    ?error,
+                    "Failed to process primary block on startup"
                 );
+                // Bring down the service as bundles processor is an essential task.
+                // TODO: more graceful shutdown.
+                return;
             }
         }
     }
 
-    while let Some(block_number) = block_imports.next().await {
-        let header = primary_chain_client
-            .header(BlockId::Number(block_number))
-            .expect("Header of imported block must exist; qed")
-            .expect("Header of imported block must exist; qed");
-        let block_info = BlockInfo {
-            hash: header.hash(),
-            parent_hash: *header.parent_hash(),
-            number: *header.number(),
-        };
-
-        if let Err(error) = block_imported(
-            primary_chain_client,
-            &processor,
-            &mut active_leaves,
-            block_info,
-        )
-        .await
-        {
-            tracing::error!(
-                target: LOG_TARGET,
-                error = ?error,
-                "Failed to process primary block"
-            );
-            break;
+    // Pause the primary block import once this channel is full.
+    let (mut block_info_sender, mut block_info_receiver) =
+        mpsc::channel(block_import_throttling_buffer_size as usize);
+
+    loop {
+        tokio::select! {
+            Some((block_number, mut block_import_throttling_sender)) = block_imports.next() => {
```

Comment:
I think it would be preferable to break out of the loop if there is no more block imports left to process.

Reply:
In which cases there are no more blocks? It should never be None in theory IMO.

Reply:
During graceful shutdown when everything is stopping